src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h

```diff
@@ -368,6 +337,87 @@ class BtreeLBAManager : public LBAManager {
     });
   }
 
+  remap_ret remap_lba_mappings(
+    Transaction &t,
+    laddr_t orig_laddr,
+    extent_len_t orig_len,
+    paddr_t orig_paddr,
+    laddr_t intermediate_base,
+    laddr_t intermediate_key,
+    std::vector<remap_entry> remaps,
+    std::vector<LogicalCachedExtentRef> extents) final {
+    LOG_PREFIX(BtreeLBAManager::remap_lba_mappings);
+    assert((intermediate_key == L_ADDR_NULL)
+      == (remaps.size() == extents.size()));
+    return seastar::do_with(
+      lba_remap_ret_t{},
+      std::move(remaps),
+      std::move(extents),
+      [&t, orig_laddr, orig_len, orig_paddr, intermediate_base, FNAME,
+      intermediate_key, this](auto &ret, auto &remaps, auto &extents) {
+      return update_refcount(t, orig_laddr, -1, false
+      ).si_then([&ret, this, &remaps, &t, orig_laddr, &extents, FNAME,
+		orig_len, orig_paddr, intermediate_key,
+		intermediate_base](auto r) {
+	ret.ruret = std::move(r.ref_update_res);
+	return trans_intr::do_for_each(
+	  boost::make_counting_iterator(size_t(0)),
+	  boost::make_counting_iterator(remaps.size()),
+	  [&remaps, &t, this, orig_laddr, orig_len, orig_paddr, &extents,
+	  intermediate_key, intermediate_base, FNAME, &ret](auto i) {
+	  auto &remap = remaps[i];
+	  auto remap_offset = remap.offset;
+	  auto remap_len = remap.len;
+	  auto remap_laddr = orig_laddr + remap_offset;
+	  auto remap_paddr = orig_paddr.add_offset(remap_offset);
+	  if (intermediate_key != L_ADDR_NULL) {
+	    remap_paddr = orig_paddr;
+	  }
+	  ceph_assert(remap_len < orig_len);
+	  ceph_assert(remap_offset + remap_len <= orig_len);
+	  ceph_assert(remap_len != 0);
+	  SUBDEBUGT(seastore_lba,
+	    "remap laddr: {}, remap paddr: {}, remap length: {}", t,
+	    remap_laddr, remap_paddr, remap_len);
+	  auto fut = alloc_extent_iertr::make_ready_future<LBAMappingRef>();
+	  auto remapped_intermediate_key = intermediate_key;
+	  if (remapped_intermediate_key != L_ADDR_NULL) {
+	    assert(intermediate_base != L_ADDR_NULL);
+	    remapped_intermediate_key += remap_offset;
+	    fut = _clone_mapping(
+	      t,
+	      remap_laddr,
+	      remap_len,
+	      remapped_intermediate_key,
+	      intermediate_base,
+	      false);
+	  } else {
+	    fut = alloc_extent(t, remap_laddr, *extents[i]);
+	  }
+	  return fut.si_then([remap_laddr, remap_len, &ret,
+			      remap_paddr](auto &&ref) {
+	    assert(ref->get_key() == remap_laddr);
+	    assert(ref->get_val() == remap_paddr);
+	    assert(ref->get_length() == remap_len);
+	    ret.remapped_mappings.emplace_back(std::move(ref));
+	    return seastar::now();
+	  });
+	});
+      }).si_then([&extents, &t, intermediate_base, intermediate_key, this] {
+	if (extents.size() > 1 && intermediate_key != L_ADDR_NULL) {
+	  return incref_extent(t, intermediate_base, extents.size() - 1
+	  ).si_then([](auto) {
+	    return seastar::now();
+	  });
+	}
+	return ref_iertr::now();
```

Comment:
Update: I think we should assert that:
* At non-clone mapping step 1, assert that the mapping is direct and ref-counter must be 1.
* At mapping with clone step 3, assert that the mapping ref-counter must be >= 1.

Reply:
> 2\. For the 2-n-th remaps, probably `alloc_extents()` is more efficient than iterative calls to `alloc_extent()` ?

`LBAManager::alloc_extents()` can only allocate laddr continuous mappings, which is not the case for the `remap` scenario.

Maybe we should add a `LBAManager::batch_modify()` interface, so that `ObjectDataHandler`'s `do_remappings`, `do_removals` and `do_insertions` can be done with one lba tree search. What do you think?

Reply:
> LBAManager::alloc_extents() can only allocate laddr continuous mappings, which is not the case for the remap scenario.

I see, remapping might need to allocate non-continuous mappings with a hole for the new write.

> Maybe we should add a LBAManager::batch_modify() interface, so that ObjectDataHandler's do_remappings, do_removals and do_insertions can be done with one lba tree search. What do you think?

I agree that the new interface is more flexible, so that update (remove+insert) the first mapping and insert the follow-up mappings can be done in the same lookup.